From bb47f2e666f31d3ebe0fa951f1d1abacc6f601b1 Mon Sep 17 00:00:00 2001
From: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Date: Thu, 6 Mar 2025 08:56:48 +0100
Subject: [PATCH 1629/1695] clk: stm32mp25: fix autonomous mode support for
 peripheral clock

GATE_<IP>AM allows to enable "Autonomous Mode" for some peripherals,
so the clock are provided in LPLV-Stop and Standby modes
(see reference manual chapter "20.4.12 Peripheral clock gating control"
Table 148.Peripheral kernel clock enabling logic).

The source clock of these clocks are the associated kernel clock as they
have the same frequency for LPLV-Stop modes and because PERxAMEN=1 need
to have also PERxEN=1 managed by the kernel clock and GATE_<IP>.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 34e37ef4e418b3ed770383ed1e2ac1c75a9abf7d clk: stm32mp25: fix autonomous mode support for peripheral clock]
Signed-off-by: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Change-Id: I1bc23ac9cd3f411b0e20e1645fddee18c714321e
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/442135
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 core/drivers/clk/clk-stm32mp25.c | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/core/drivers/clk/clk-stm32mp25.c b/core/drivers/clk/clk-stm32mp25.c
index b50b9cf25..1bc36df25 100644
--- a/core/drivers/clk/clk-stm32mp25.c
+++ b/core/drivers/clk/clk-stm32mp25.c
@@ -3877,11 +3877,11 @@ static RIF_GATE(ck_icn_p_hpdma3, &ck_icn_ls_mcu, 0, GATE_HPDMA3,
 		RCC_RIF_HPDMA3);
 static RIF_GATE(ck_icn_p_lpdma, &ck_icn_ls_mcu, 0, GATE_LPDMA,
 		RCC_RIF_LPDMA);
-static RIF_GATE(ck_icn_p_lpdma_am, &ck_msi_ker, 0, GATE_LPDMAAM,
+static RIF_GATE(ck_icn_p_lpdma_am, &ck_icn_p_lpdma, 0, GATE_LPDMAAM,
 		RCC_RIF_LPDMA);
 static RIF_GATE(ck_icn_p_ipcc1, &ck_icn_ls_mcu, 0, GATE_IPCC1, RCC_RIF_IPCC1);
 static RIF_GATE(ck_icn_p_ipcc2, &ck_icn_ls_mcu, 0, GATE_IPCC2, RCC_RIF_IPCC2);
-static RIF_GATE(ck_icn_p_ipcc2_am, &ck_icn_ls_mcu, 0, GATE_IPCC2AM,
+static RIF_GATE(ck_icn_p_ipcc2_am, &ck_icn_p_ipcc2, 0, GATE_IPCC2AM,
 		RCC_RIF_IPCC2);
 static RIF_GATE(ck_icn_p_hsem, &ck_icn_ls_mcu, 0, GATE_HSEM, RCC_RIF_HSEM);
 static RIF_GATE(ck_icn_p_gpioa, &ck_icn_ls_mcu, 0, GATE_GPIOA, RCC_RIF_GPIOA);
@@ -3896,9 +3896,11 @@ static RIF_GATE(ck_icn_p_gpioi, &ck_icn_ls_mcu, 0, GATE_GPIOI, RCC_RIF_GPIOI);
 static RIF_GATE(ck_icn_p_gpioj, &ck_icn_ls_mcu, 0, GATE_GPIOJ, RCC_RIF_GPIOJ);
 static RIF_GATE(ck_icn_p_gpiok, &ck_icn_ls_mcu, 0, GATE_GPIOK, RCC_RIF_GPIOK);
 static RIF_GATE(ck_icn_p_gpioz, &ck_icn_ls_mcu, 0, GATE_GPIOZ, RCC_RIF_GPIOZ);
-static RIF_GATE(ck_icn_p_gpioz_am, &ck_msi_ker, 0, GATE_GPIOZAM, RCC_RIF_GPIOZ);
+static RIF_GATE(ck_icn_p_gpioz_am, &ck_icn_p_gpioz, 0, GATE_GPIOZAM,
+		RCC_RIF_GPIOZ);
 static RIF_GATE(ck_icn_p_rtc, &ck_icn_ls_mcu, 0, GATE_RTC, RCC_RIF_RTC_TAMP);
-static RIF_GATE(ck_icn_p_rtc_am, &ck_msi_ker, 0, GATE_RTCAM, RCC_RIF_RTC_TAMP);
+static RIF_GATE(ck_icn_p_rtc_am, &ck_icn_p_rtc, 0, GATE_RTCAM,
+		RCC_RIF_RTC_TAMP);
 static RIF_COMPOSITE(ck_rtc, 4, PARENT(&ck_off, &ck_lse, &ck_lsi, &ck_hse_rtc),
 		     0, GATE_RTCCK, NO_DIV, MUX_RTC, RCC_RIF_RTC_TAMP);
 static RIF_GATE(ck_icn_p_bsec, &ck_icn_apb3, 0, GATE_BSEC, RCC_RIF_BSEC);
-- 
2.47.2

